pkg/utils/cgroup/pod.go

```diff
@@ -0,0 +1,72 @@
+package cgroup
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	utilsio "github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pkg/utils/io"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+type PodCGroupSlice struct {
+	KubepodsSlicePath string
+}
+
+func (cg PodCGroupSlice) FindPodDir(podUID string) (int, error) {
+	if cg.KubepodsSlicePath == "" {
+		return -1, errors.New("no cgroup detected")
+	}
+
+	// /sys/fs/cgroup/blkio/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-podaadcc749_6776_4933_990d_d50f260f5d46.slice/blkio.throttle.write_bps_device
+	podUID = strings.ReplaceAll(podUID, "-", "_")
+	for _, p := range []string{
+		"%s/kubepods-besteffort.slice/kubepods-besteffort-pod%s.slice",
+		"%s/kubepods-burstable.slice/kubepods-burstable-pod%s.slice",
+		"%s/kubepods-pod%s.slice",
+	} {
+		p = fmt.Sprintf(p, cg.KubepodsSlicePath, podUID)
+		fd, err := unix.Open(p, utilsio.O_PATH, 0)
+		if err == nil {
+			return fd, nil
+		}
+		if !errors.Is(err, fs.ErrNotExist) {
+			return -1, fmt.Errorf("failed to open CGroup path %q: %w", p, err)
+		}
+	}
+	return -1, errors.New("pod CGroup path not found")
+}
+
+// DetectCgroupVersion detects the cgroup version of the cgroup mounted at cgroupMountPath.
+// It only cares about kubepods.slice
+//
+// Takes the path to the cgroup mount point as an argument. Usually /sys/fs/cgroup.
+// Returns the path to kubepods.slice, the IO implementation for the detected version.
+func DetectCgroupVersion(cgroupMountPath string) (PodCGroupSlice, IO, error) {
+	v2Path := filepath.Join(cgroupMountPath, "kubepods.slice")
+	_, err := os.Stat(v2Path + "/io.max")
+	if err == nil {
+		log.Infof("Detected cgroup v2 at %s", v2Path)
+		return PodCGroupSlice{v2Path}, V2{}, nil
+	}
+	if !errors.Is(err, os.ErrNotExist) {
+		return PodCGroupSlice{}, nil, err
+	}
+
+	v1Path := filepath.Join(cgroupMountPath, "blkio/kubepods.slice")
```

Comment:
Do we need to create a directory in the codes? The v1 cgroup path was previously established in the YAML.

Reply:
No. Each directory in cgroupfs corresponds to a cgroup. kubelet is responsible for creating the cgroup. Even we create the cgroup, we cannot put processes into this cgroup.

Reply:
okay,  you had tested in new cluster right?

Reply:
I've checked the log of our CI, which creates new cluster at each run.
```
2024-06-11T13:15:22.367483041+08:00 stderr F time="2024-06-11T13:15:22+08:00" level=info msg="Detected cgroup v1 at /host/sys/fs/cgroup/blkio/kubepods.slice"
```
And I think we don't need to worry about this, because csi-plugin itself is a pod, it is also in the cgroup kubepods.slice. So it must exist when csi-plugin is initializing.

Reply:
okay